filtering.py

```diff
@@ -0,0 +1,561 @@
+        estimated spatial covariance matrices
+
+
+    Note
+    -----
+        * You need an initial estimate for the sources to apply this
+          algorithm. This is precisely what the :func:`wiener` function does.
+
+        * This algorithm *is not* an implementation of the "exact" EM
+          proposed in [1]_. In particular, it does compute the posterior
+          covariance matrices the same (exact) way. Instead, it uses the
+          simplified approximate scheme initially proposed in [5]_ and further
+          refined in [3]_, [4]_, that boils down to just take the empirical
+          covariance of the recent source estimates, followed by a weighted
+          average for the update of the spatial covariance matrix. It has been
+          empirically demonstrated that this simplified algorithm is more
+          robust for music separation.
+
+    Warning
+    -------
+        It is *very* important to make sure `x.dtype` is `torch.float64`
+        if you want double precision, because this function will **not**
+        do such conversion for you from `torch.complex32`, in case you want the
+        smaller RAM usage on purpose.
+
+        It is usually always better in terms of quality to have double
+        precision, by e.g. calling :func:`expectation_maximization`
+        with ``x.to(torch.float64)``.
+
+    """
+    # to avoid dividing by zero
+    if eps is None:
+        eps = torch.tensor(torch.finfo(x.dtype).eps,
+                           dtype=x.dtype, device=x.device)
+
+    # dimensions
+    (nb_frames, nb_bins, nb_channels) = x.shape[:-1]
+    nb_sources = y.shape[-1]
+
+    if verbose:
+        print('Number of iterations: ', iterations)
+
+    regularization = torch.cat(
+        (
+            torch.eye(2, dtype=x.dtype, device=x.device)[..., None],
+            torch.zeros((2, 2, 1), dtype=x.dtype, device=x.device)
+        ),
+        dim=2
+    )
+    regularization = torch.sqrt(eps)*(
+        regularization[None, None, ...].expand(
+                            (-1, nb_bins, -1, -1, -1))
+    )
+
+    # allocate the spatial covariance matrices
+    R = [
+        torch.zeros(
+            (nb_bins, nb_channels, nb_channels, 2),
+            dtype=x.dtype, device=x.device
+        )
+        for j in range(nb_sources)
+    ]
+    weight = torch.zeros(
+        (nb_bins,),
+        dtype=x.dtype, device=x.device
+    )
+
+    for it in range(iterations):
+        # constructing the mixture covariance matrix. Doing it with a loop
+        # to avoid storing anytime in RAM the whole 6D tensor
+        if verbose:
+            print('EM, iteration', (it+1))
+
+        # update the PSD as the average spectrogram over channels
+        v = torch.mean(
+            torch.abs(y[..., 0, :])**2 + torch.abs(y[..., 1, :])**2,
+            dim=-2
+        )
+
+        # update spatial covariance matrices (weighted update)
+        batch_size=200
```

Comment:
magic number?

Reply:
yes

Reply:
not important though, any number works the same

Reply:
lets make it a function parameter, just in case

Reply:
ok

Reply:
ok

Reply:
done